File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
function solution() {
  const fs = require("fs");
  const filePath =
    process.platform === "linux" ? "/dev/stdin" : "./example.txt";
  const input = fs
    .readFileSync(filePath)
    .toString()
    .trim()
    .split("\n")
    .map((v) => v.replace("\r", ""));
  const [n, m] = input.shift().split(" ").map(Number);
  const visited = Array.from(Array(n), () => Array(m).fill(false));
  const dx = [0, -1, 0, 1];
  const dy = [1, 0, -1, 0];
  const queue = [[0, 0, 1]];
  visited[0][0] = true;

  while (queue.length) {
    const [cx, cy, cnt] = queue.shift();
    if (cx === n - 1 && cy === m - 1) {
      console.log(cnt);
      break;
    }

    for (let i = 0; i < 4; i++) {
      const nx = cx + dx[i];
      const ny = cy + dy[i];

      if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
      if (visited[nx][ny]) continue;
      if (input[nx][ny] === "0") continue;

      visited[nx][ny] = true;
      queue.push([nx, ny, cnt + 1]);
    }
  }
}

solution();

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque


def input():
    return sys.stdin.readline().rstrip()


def solution():
    N, M = map(int, input().split())
    board = [list(map(int, input())) for _ in range(N)]
    dist = [[1 for _ in range(M)] for _ in range(N)]
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]

    def bfs():
        queue = deque()
        queue.append((0, 0))

        while queue:
            x, y = queue.popleft()

            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]

                if nx < 0 or nx >= N or ny < 0 or ny >= M:
                    continue
                if board[nx][ny] == 0:
                    continue
                if dist[nx][ny] != 1:
                    continue

                board[nx][ny] = True
                dist[nx][ny] = dist[x][y] + 1
                queue.append((nx, ny))

        return

    bfs()

    return dist[N - 1][M - 1]


print(solution())
